# LAB 04 -- WEEK 05 -- Hypothesis Testing 


## Challenge 1: Pokemon Dataset

In [1]:
import pandas as pd
import scipy.stats as st
import numpy as np

# Load Pokemon CSV 
df_poke = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/pokemon.csv")

### Hypothesis 1: Dragon vs. Grass HP

H1: Dragon type Pokemons have greater HP than Grass type.

Test: Independent T-Test (One-tailed)


In [2]:
# T-Test for Dragon vs. Grass types 

# Get HP data for Dragon and Grass types (use main 'Type 1' bc many don't have 2nd type)
dragon_hp = df_poke[df_poke['Type 1'] == 'Dragon']['HP']
grass_hp = df_poke[df_poke['Type 1'] == 'Grass']['HP']

# Independent T-test (alternative='greater' to check if Dragon > Grass)
stat, p_val = st.ttest_ind(dragon_hp, grass_hp, alternative='greater')

print(f"Dragon vs. Grass HP -- p-Value: {p_val}")

Dragon vs. Grass HP -- p-Value: 0.0002567969150153481


In [ ]:
# Print T-test Results

thresh = 0.05
if p_val < thresh:
    print("Conclusion: H0 can be rejected.")
    print("Result: With a significance threshold of 5 percent, Dragon type Pokemon have greater HP on average than the Grass type.")
else:
    print("\nConclusion: Failed to reject H0!")
    print("Result: No statistical evidence found to reject the Null hypothesis.")

Conclusion: H0 can be rejected.
Result: With a significance threshold of 5 percent, Dragon type Pokemon have greater HP on average than the Grass type.


### Hypothesis 2: Legendary vs. Non-Legendary

H1: Legendary Pokemon have fundamentally different strengths compared to Non-Legendary ones.

Test: Independent T-Test (Two-tailed) 

In [ ]:
# Get p Values for all relevant Stats for the Legendary and Non-L. Pokemon groups

# Separate into two dfs
legendary = df_poke[df_poke['Legendary'] == True]
non_legendary = df_poke[df_poke['Legendary'] == False]

# Loop through the Attack & Defense stats
stats_to_test = ['HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

for stat_name in stats_to_test:
    t_stat, p_val = st.ttest_ind(legendary[stat_name], non_legendary[stat_name], alternative='two-sided')
    # alternative='two-sided' -- check for any deviation to the right or left of zero +/-
    print(f"{stat_name} p-value: {p_val}")

HP p-value: 3.3306476848461913e-15
Attack p-value: 7.827253003205333e-24
Defense p-value: 1.5842226094427255e-12
Sp. Atk p-value: 6.314915770427265e-41
Sp. Def p-value: 1.8439809580409597e-26
Speed p-value: 2.3540754436898437e-21


In [5]:
# Print Findings 

print("Conclusion: As the Legendary Pokemons have p-values *far below* 0.05 in all relevant attack and defense variables, H0 can be summarily rejected.")
print("Result: Legendary Pokemons are fundamentally different, and display much greater strengths overall across all characteristics.")

Conclusion: As the Legendary Pokemons have p-values *far below* 0.05 in all relevant attack and defense variables, H0 can be summarily rejected.
Result: Legendary Pokemons are fundamentally different, and display much greater strengths overall across all characteristics.


## Challenge 2: California Housing Dataset 

In [6]:
import pandas as pd
import scipy.stats as st
import numpy as np

# Load Housing CSV 
df_house = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/california_housing.csv")

### Hypothesis 1: Closeness to schools or hospitals increases

H1:  Houses closer to a school or hospital have greater median value.

Test: Independent T-Test (One-tailed) 

In [ ]:
# Calculate Euclidean Distance to Fixed School & Hospital Coordinates
# Return True if closer than 0.5 coordinate units to either

# Define coordinates
school = np.array([-118, 34])
hospital = np.array([-122, 37])

# Calculate Euclidean distance with np.linalg.norm-- Return True if is_close < 0.5
def is_close(row):
    house = np.array([row['longitude'], row['latitude']])  # Get house coordinates 
    dist_school = np.linalg.norm(house - school)
    dist_hospital = np.linalg.norm(house - hospital)
    return ((dist_school < 0.50) or (dist_hospital < 0.50)) # Return Bool if either is smaller than 0.5 

# Apply closeness function and store to new Boolean column 'is_close' 
df_house['is_close'] = df_house.apply(is_close, axis=1)

# Separate median house prices into two groups: close or far 
close_prices = df_house[df_house['is_close'] == True]['median_house_value']
far_prices = df_house[df_house['is_close'] == False]['median_house_value']


In [ ]:
# Independent T-test 

# alternative='greater' -- check if Close > Far; Use equal_var=False to minimize variance in Bin sizes
t_stat, p_val = st.ttest_ind(close_prices, far_prices, alternative='greater', equal_var=False)

print(f"Close vs. Far Housing Prices -- p-Value: {p_val}")


Close vs. Far Housing Prices -- p-Value: 1.5032478884296307e-301


# Findings & Results

With a negative power of -301, the p-Value is so far below the 5 percent threshold that the Null Hypothesis must be rejected. In this dataset at least, the Euclidean distance to schools and hospitals is very strongly tied to the market value of the house. So real estate within a 0.5 coordinate radius of schools or hospitals is generally more expensive than properties that are further away. Even in a rich state with strong infrastructure like California, the access to basic public services is a central and decisive factor determining a house's value.   